### PR TITLE
Accumulators

### PR DESCRIPTION
**Pull request by Mati-AAU**

### CODE REVIEW

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/MaskedAccumulationStage.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        glActiveTexture(inputFramebuffer.textureUnitPair.textureUnit)
+        glBindTexture(GLES20.GL_TEXTURE_2D, inputFramebuffer.textureHandle)
+
+        // Input old accumulator framebuffer
+        val oldAccumulatorHandle = glGetUniformLocation(program, "oldAccumulator")
+        glUniform1i(oldAccumulatorHandle, oldAccumulator.textureUnitPair.textureUnitIndex)
+        glActiveTexture(oldAccumulator.textureUnitPair.textureUnit)
+        glBindTexture(GLES20.GL_TEXTURE_2D, oldAccumulator.textureHandle)
+
+        // Input old accumulator framebuffer
```
Comment (Casper-NS) [inline app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/MaskedAccumulationStage.kt:47]: Fake news!?!?!

File: app/src/main/res/raw/accumulation_shader.glsl [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+uniform sampler2D oldAccumulator;
+
+void main() {
+    vec2 samplersUV = gl_FragCoord.xy / resolution;
+
+    vec4 col1 = texture2D(framebuffer, samplersUV);
+    vec4 col2 = texture2D(oldAccumulator, samplersUV);
+
+
+    float add = col2.x + 0.2 * col1.x - 0.2 * (1.0-col1.x);
```
Comment (Casper-NS) [inline app/src/main/res/raw/accumulation_shader.glsl:16]: Expose the accumalotor value

File: app/src/main/res/raw/step_shader.glsl [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+uniform vec2 resolution;
+uniform sampler2D framebuffer;
+
+void main() {
+    vec2 uv = gl_FragCoord.xy / resolution;
+
+    vec4 color = texture2D(framebuffer, uv);
+
+    float value = 1.0 - step(0.5, color.a);
```
Comment (Casper-NS) [inline app/src/main/res/raw/step_shader.glsl:12]: Expose the threshold for the step

File: app/src/main/res/raw/masked_accumulation_shader.glsl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    vec4 col1 = texture2D(framebuffer, samplersUV);
+    vec4 col2 = texture2D(oldAccumulator, samplersUV);
+    vec4 mask = texture2D(mask, samplersUV);
+
+    float binaryInput = step(0.99, col1.x);
+    float maskValue = step(0.99, mask.x);
+
+    float add = (col2.a + (maskValue) * (0.2 * binaryInput - 0.2 * (1.0-binaryInput)));
+    //float add = step(0.99, mask.x);
```
Comment (Casper-NS) [inline app/src/main/res/raw/masked_accumulation_shader.glsl:21]: We dont like commented code.

File: app/src/main/res/raw/segment_postprocess_shader.glsl [hunk 5]
```diff
@@ -7,7 +7,7 @@ uniform sampler2D framebuffer;
 void main() {
     vec2 uv = gl_FragCoord.xy / resolution;
```
Comment (Casper-NS) [inline app/src/main/res/raw/segment_postprocess_shader.glsl:9]: Simplify and add a comment explaining this.

**Review by Casper-NS — CHANGES_REQUESTED** [review]
This PR is cursed, plz fix

File: app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/MaskedAccumulationStage.kt (hunk 1 above)
Comment (Mati-AAU) [reply app/src/main/java/dk/scuffed/whiteboardapp/pipeline/stages/opengl_process_stages/MaskedAccumulationStage.kt]: done

File: app/src/main/res/raw/accumulation_shader.glsl (hunk 2 above)
Comment (Mati-AAU) [reply app/src/main/res/raw/accumulation_shader.glsl]: done

File: app/src/main/res/raw/step_shader.glsl (hunk 3 above)
Comment (Mati-AAU) [reply app/src/main/res/raw/step_shader.glsl]: done

File: app/src/main/res/raw/masked_accumulation_shader.glsl (hunk 4 above)
Comment (Mati-AAU) [reply app/src/main/res/raw/masked_accumulation_shader.glsl]: done

File: app/src/main/res/raw/masked_accumulation_shader.glsl [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+uniform sampler2D oldAccumulator;
+uniform sampler2D mask;
+
+void main() {
+    vec2 samplersUV = gl_FragCoord.xy / resolution;
+
+    vec4 col1 = texture2D(framebuffer, samplersUV);
+    vec4 col2 = texture2D(oldAccumulator, samplersUV);
+    vec4 mask = texture2D(mask, samplersUV);
+
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply app/src/main/res/raw/masked_accumulation_shader.glsl:18]: fixed

File: app/src/main/res/raw/masking_shader.glsl [hunk 7]
```diff
@@ -17,7 +17,7 @@ void main() {
     vec4 mask = texture2D(mask, samplersUV);
 
     // annoying hack because the segmentor outputs a slightly blue color for no reason at all.
-    float actualmask = step(0.1, mask.x + mask.y + mask.z);
+    float actualmask = step(0.99, mask.x + mask.y + mask.z);
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply app/src/main/res/raw/masking_shader.glsl]: done

File: app/src/main/res/raw/segment_postprocess_shader.glsl (hunk 5 above)
Comment (Mati-AAU) [reply app/src/main/res/raw/segment_postprocess_shader.glsl:9]: fix

Comment (Mati-AAU) [reply app/src/main/res/raw/segment_postprocess_shader.glsl:9]: ed

File: app/src/main/res/raw/segmentation_accumulation_shader.glsl [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    vec2 samplersUV = gl_FragCoord.xy / resolution;
+
+    vec4 col1 = texture2D(segmentation, samplersUV);
+    vec4 col2 = texture2D(oldAccumulator, samplersUV);
+
+
+    float segmentMap = 1.0 - step(0.05, col1.z);
+
+
+    float add = (col2.x + 0.1) * segmentMap;
```
[the remark this answers is not in the document]
Comment (Mati-AAU) [reply app/src/main/res/raw/segmentation_accumulation_shader.glsl]: good things :)

**Review by Casper-NS — APPROVED** [review]
lgtm is less cursed now

**Pull request merged**